crates/ty_python_semantic/resources/mdtest/narrow/len.md

```diff
@@ -0,0 +1,128 @@
+# Narrowing for `len(..)` checks
+
+When `len(x)` is used in a boolean context, we can narrow the type of `x` based on whether `len(x)`
+is truthy (non-zero) or falsy (zero).
+
+We apply `~AlwaysFalsy` narrowing when ANY part of the type is narrowable (string/bytes literals,
+`LiteralString`, tuples). This removes types that are always falsy (like `Literal[""]`) while
+leaving non-narrowable types (like `str`, `list`) unchanged.
+
+## String literals
+
+The intersection with `~AlwaysFalsy` simplifies to just the non-empty literal.
+
+```py
+from typing import Literal
+
+def _(x: Literal["foo", ""]):
+    if len(x):
+        reveal_type(x)  # revealed: Literal["foo"]
+    else:
+        reveal_type(x)  # revealed: Literal[""]
+```
+
+## Bytes literals
+
+```py
+from typing import Literal
+
+def _(x: Literal[b"foo", b""]):
+    if len(x):
+        reveal_type(x)  # revealed: Literal[b"foo"]
+    else:
+        reveal_type(x)  # revealed: Literal[b""]
+```
+
+## LiteralString
+
+```toml
+[environment]
+python-version = "3.11"
+```
+
+```py
+from typing import LiteralString
+
+def _(x: LiteralString):
+    if len(x):
+        reveal_type(x)  # revealed: LiteralString & ~Literal[""]
+    else:
+        reveal_type(x)  # revealed: Literal[""]
+```
+
+## Tuples
+
+```py
+def _(x: tuple[int, ...]):
+    if len(x):
+        reveal_type(x)  # revealed: tuple[int, ...] & ~AlwaysFalsy
+    else:
+        reveal_type(x)  # revealed: tuple[int, ...] & ~AlwaysTruthy
```

Comment:
you could _possibly_ add TODOs here that ideally we'd narrow these types further, e.g. to

```suggestion
    if len(x):
        reveal_type(x)  # revealed: tuple[int, ...] & ~tuple[()]
    else:
        reveal_type(x)  # revealed: tuple[()]
```

(which is related to https://github.com/astral-sh/ty/issues/560)